### PR TITLE
add Policy_gradient pytorch

### PR DESCRIPTION
**Pull request by Asugawara**
Hi! @lanctot

I know it's been a while since last my Pull Request.
I've tried to implement loss functions and Policy Gradient Agent in PyTorch!
Could you pleased check it out?

Below figures are a comparison of the Exploitability averaged over four times of Policy_gradient(A2C) agent.
(I've also fixed the broken links.)

スクリーンショット 2021-01-24 16 38 32
スクリーンショット 2021-01-24 16 38 41

### CODE REVIEW

**Comment by lanctot** [comment]
Thanks @Asugawara, this is great! We will take a look in the next week or so.

**Comment by lanctot** [comment]
Hi @Asugawara, can you try running A2C on Catch to see if it solves it?

**Comment by Asugawara** [comment]
Yes! Thank you for your feedback!

**Comment by Asugawara** [comment]
Hi @lanctot !
let me ask a question.
Below figure is a comparison of "the average returns" averaged over 3 times of TensorFlow agents(DQN and A2C).
Is A2C supposed to increase average returns like DQN?
スクリーンショット 2021-01-26 6 34 55


if TF's A2C would need to fix, PyTorch’s A2C, too!!
スクリーンショット 2021-01-26 6 35 18

**Comment by lanctot** [comment]
Yeah this is not the first time I'm seeing this. I think the hyper-parameters we chose as defaults were based on the runs we did for Kuhn/Leduc poker and turn out to be bad for catch. I will ask around for good hyper-parameters and get back to you.

**Comment by Asugawara** [comment]
> I will ask around for good hyper-parameters and get back to you

Thank you for your concern! Dr. @lanctot 
I tried to run with other hyper-parameters, and I found a good combination!
スクリーンショット 2021-01-26 20 04 52

**Comment by lanctot** [comment]
@Asugawara this is excellent! Thank you.

Can I ask you to add two examples (both on Catch):

1. Using the Tensforflow A2C (similar to python/examples/kuhn_policy_gradient.py, except catch_tf_policy_gradient.py)
2. The same example with pythorch (python/examples/catch_pytorch_policy_gradient.py)

Which uses the hyper-parameters you found (that would lead to those two curves)?

This would be very helpful for us and future users looking at using policy gradient for single-agent RL tasks.

**Comment by lanctot** [comment]
Note that I'm not asking for the examples to produce the actual curves, so no need for them to use any graphing libraries.. I just meant as example similar to the other ones. If you prefer colabs, that's great too. :)

**Comment by Asugawara** [comment]
Of Course!

>1.Using the Tensforflow A2C (similar to python/examples/kuhn_policy_gradient.py, except catch_tf_policy_gradient.py)

There is already a file that uses single agent algorithms on Catch.
So, I only changed A2C hyper-parameter.
https://github.com/deepmind/open_spiel/blob/daf0c353f70ccee003f63f09d0d0314ddb516295/open_spiel/python/examples/single_agent_catch.py

>2.The same example with pytorch (python/examples/catch_pytorch_policy_gradient.py)
>no need for them to use any graphing libraries..

I added!
https://github.com/deepmind/open_spiel/blob/daf0c353f70ccee003f63f09d0d0314ddb516295/open_spiel/python/examples/catch_pytorch_policy_gradient.py

>This would be very helpful for us and future users looking at using policy gradient for single-agent RL tasks.

Great!

**Comment by lanctot** [comment]
Thanks @Asugawara, I did not realize we did put a single_agent catch example. Now I'm slightly worried about changing the hyper-parameters (in the original one) because we may have chosen the hyper-parameters that are the best for all of { "a2c", "rpg", "rmpg", "qpg" } rather than just A2C.

Can I ask you to run your TF on catch using those older hyper-parameters?

```
          hidden_layers_sizes=[32, 32],
          batch_size=32,
          entropy_cost=0.001,
          critic_learning_rate=0.01,
          pi_learning_rate=0.01,
          num_critic_before_pi=5)
```

Do the results look the same?

(Apologies, I promise it will be the last request!)

**Comment by Asugawara** [comment]
Thank you Dr. @lanctot!!

> Now I'm slightly worried about changing the hyper-parameters

I understand how your concern!

>Can I ask you to run your TF on catch using those older hyper-parameters?

[The first graph on Catch](https://github.com/deepmind/open_spiel/pull/491#issuecomment-767132318) was the result of a test using this parameter!
It didn't go so well.

>(Apologies, I promise it will be the last request!)

No problem! I'll test it as many times as you want.

**Comment by lanctot** [comment]
Really? 

@fzvinicius did we run these in catch or just copied them from the Kuhn/Leduc experiments? What is your opinion.. should we stick with the new ones?

I also asked another colleague too, Vlad Mnih, he thought those original ones should also work. When I look at those hyper-params I am surprised that they don't work. Do you know if it was using sgd or adam?

**Comment by lanctot** [comment]
Are you sure it was not the defaults copies from kuhn_policy_gradient.py example?

**Comment by lanctot** [comment]
Ok I will also run it on my side today too to verify

**Comment by Asugawara** [comment]
Yes, Now I run single_agent_catch.py! (I only change eval_interval!)
>Do you know if it was using sgd or adam?

[it's use sgd.](https://github.com/deepmind/open_spiel/blob/master/open_spiel/python/algorithms/policy_gradient.py#L113)
```
python single_agent_catch.py --algorithm a2c --num_episodes 100000
I0128 20:46:51.071910 4796661184 single_agent_catch.py:117] --------------------------------------------------------------------------------
I0128 20:46:51.072028 4796661184 single_agent_catch.py:118] Episode 10000
I0128 20:46:51.072077 4796661184 single_agent_catch.py:119] Loss: (1.2253155, 0.50992644)
I0128 20:46:51.294698 4796661184 single_agent_catch.py:121] Avg return: -0.54
I0128 20:47:15.846596 4796661184 single_agent_catch.py:117] --------------------------------------------------------------------------------
I0128 20:47:15.846687 4796661184 single_agent_catch.py:118] Episode 20000
I0128 20:47:15.846734 4796661184 single_agent_catch.py:119] Loss: (0.7215541, 0.11666449)
I0128 20:47:16.069784 4796661184 single_agent_catch.py:121] Avg return: -0.48
I0128 20:47:40.604400 4796661184 single_agent_catch.py:117] --------------------------------------------------------------------------------
I0128 20:47:40.604491 4796661184 single_agent_catch.py:118] Episode 30000
I0128 20:47:40.604538 4796661184 single_agent_catch.py:119] Loss: (0.7172244, -0.05672247)
I0128 20:47:40.826575 4796661184 single_agent_catch.py:121] Avg return: -0.64
I0128 20:48:05.259879 4796661184 single_agent_catch.py:117] --------------------------------------------------------------------------------
I0128 20:48:05.259968 4796661184 single_agent_catch.py:118] Episode 40000
I0128 20:48:05.260014 4796661184 single_agent_catch.py:119] Loss: (0.72937214, -0.0034824503)
I0128 20:48:05.484017 4796661184 single_agent_catch.py:121] Avg return: -0.64
I0128 20:48:29.929751 4796661184 single_agent_catch.py:117] --------------------------------------------------------------------------------
I0128 20:48:29.929840 4796661184 single_agent_catch.py:118] Episode 50000
I0128 20:48:29.929888 4796661184 single_agent_catch.py:119] Loss: (0.10892948, -0.3476112)
I0128 20:48:30.152997 4796661184 single_agent_catch.py:121] Avg return: -0.5
I0128 20:48:54.711845 4796661184 single_agent_catch.py:117] --------------------------------------------------------------------------------
I0128 20:48:54.711933 4796661184 single_agent_catch.py:118] Episode 60000
I0128 20:48:54.711980 4796661184 single_agent_catch.py:119] Loss: (0.26016766, -0.50990385)
I0128 20:48:54.936377 4796661184 single_agent_catch.py:121] Avg return: -0.46
I0128 20:49:19.748204 4796661184 single_agent_catch.py:117] --------------------------------------------------------------------------------
I0128 20:49:19.748297 4796661184 single_agent_catch.py:118] Episode 70000
I0128 20:49:19.748346 4796661184 single_agent_catch.py:119] Loss: (1.1228546, 0.5500473)
I0128 20:49:19.972973 4796661184 single_agent_catch.py:121] Avg return: -0.56
I0128 20:49:44.719754 4796661184 single_agent_catch.py:117] --------------------------------------------------------------------------------
I0128 20:49:44.719845 4796661184 single_agent_catch.py:118] Episode 80000
I0128 20:49:44.719889 4796661184 single_agent_catch.py:119] Loss: (0.26193744, -0.47332555)
I0128 20:49:44.941900 4796661184 single_agent_catch.py:121] Avg return: -0.52
I0128 20:50:09.620074 4796661184 single_agent_catch.py:117] --------------------------------------------------------------------------------
I0128 20:50:09.620164 4796661184 single_agent_catch.py:118] Episode 90000
I0128 20:50:09.620210 4796661184 single_agent_catch.py:119] Loss: (0.7070945, 0.17234686)
I0128 20:50:09.843639 4796661184 single_agent_catch.py:121] Avg return: -0.58
```

**Comment by Asugawara** [comment]
>Are you sure it was not the defaults copied from kuhn_policy_gradient.py example?

No, I use single_agent_catch.py 's hyper-parameter!

**Comment by Asugawara** [comment]
>Ok I will also run it on my side today too to verify

Thanks a lot!
I'm looking forward to your verification and your colleagues' reply!

**Comment by lanctot** [comment]
@Asugawara, indeed I get the same behavior with those original parameters. I will now try the new ones.

**Comment by lanctot** [comment]
@Asugawara I can confirm the new values do seem to be learning. @fzvinicius I plan to keep these new parameters, please let me know what you think.

**Comment by lanctot** [comment]
Thanks @fzvinicius for the internal review!

He made a good point: he thinks we changed the policy gradient code to use sgd by default instead of adam, and we forgot to update these examples. 

I've imported it now, will get merged on Monday. Thanks again for doing this. Nice to see our pytorch algorithms growing!

**Comment by lanctot** [comment]
(We also changed the learning rate of DQN to 0.1 and it solves it in about ~3k steps. So possibly higher learning rates or num_critic_before_pi=1 could make the PG ones even faster.)

**Comment by Asugawara** [comment]
Thank you for your verification! Dr. @lanctot !!

>He made a good point: he thinks we changed the policy gradient code to use sgd by default instead of adam, and we forgot to update these examples.

Well, that certainly seems likely!
